### ISSUE TITLE
setup.cfg vs setup.py

### ISSUE THREAD

**Issue by dstufft** [pypa/packaging-problems#29]
What should you do in setup.cfg vs what should you do in setup.py? This is made even more difficult to figure out by the distutils2 examples floating out there and projects like d2to1.

**Comment by brainwane** [comment]
I'm going to move into this discussion this question https://github.com/pypa/twine/issues/258#issuecomment-355697656 from @lrntct :

> I just ran into the issue. I had the URL in setup.cfg instead of setup.py. Is it an expected behaviour to ignore the URL when defined in setup.cfg? Maybe related: pypa/setuptools#862

**Comment by brainwane** [comment]
https://packaging.python.org/tutorials/distributing-packages/#initial-files has some guidance but could have more.

**Comment by brainwane** [comment]
I would also welcome clarification on how `pyproject.toml` and [PEP 518](https://www.python.org/dev/peps/pep-0518/) figure into this.

**Comment by njsmith** [comment]
Historically, `setup.cfg` is "some file that distutils would parse using some code", so the behavior was pretty unspecified. I think in practice it uses the Python stdlib's configparser package, which is a limited and quirky format (so e.g., you can't necessarily parse it from other languages and get the same result, or put anything more interesting than some strings in there), and it ignores unrecognized sections, so other tools like flake8 and tox started optionally sticking their configuration in there too.

`pyproject.toml` is the new thing, and it has two major differences: (1) it uses a more powerful and better documented language (TOML), (2) it's defined by a general Python-level standard instead of being something specific to setup.py/distutils – and in fact one of its major purposes is to be something pip can look at *before* invoking `setup.py` (as in PEP 518), or can look at and then decide *not* to invoke `setup.py` at all ([PEP 517](https://www.python.org/dev/peps/pep-0517/)). (In fact, part of the motivation here is to give us an escape hatch from the decade+ of messy stuff that has accumulated around distutils.) And since it's going to be mandatory for lots of python projects and has a nice format, PEP 518 also provides a standard way for tools to start sticking their configuration in there, if they want to. I'm not aware of any tools that are doing this yet besides [towncrier](https://github.com/hawkowl/towncrier), probably more will start once pip 10 ships and `pyproject.toml` starts being a real thing you can use instead of just a cool idea in some standards docs.

**Comment by Ivoz** [comment]
A major problem is that `setup.cfg` has in the past been used to configure how to *build* a project, from both a user and maintainer perspective, but more recently in history has been cooped by other tools to be used as a way to *configure* / describe a project purely from a maintainer perspective (e.g pbr). So it has been pulled in two directions, which once every so often collide and create confusion for new users / packagers. A single file being used for multiple purposes, for different things, with documentation coming from different places... is not a great situation to be in. Hence I think it should be moved away from in general in favour of other solutions without double-uses, like the new toml standard.

**Comment by ncoghlan** [comment]
Yeah, that's exactly the direction we're going. While this confusion will likely always exist for setuptools/distutils users for historical reasons, for future users we should reach a point where they only need pyproject.toml plus their preferred packaging tool's input format (e.g. "flit.ini" for folks using `flit`)
